src/llvm-late-gc-lowering.cpp

```diff
@@ -0,0 +1,1351 @@
+        dbgs() << "\n\tUpExposedUses: ";
+        dumpBitVectorValues(S, S.BBStates[&BB].UpExposedUses);
+        dbgs() << "\n\tDownExposedUnrooted: ";
+        dumpBitVectorValues(S, S.BBStates[&BB].DownExposedUnrooted);
+        dbgs() << "\n\tLiveIn: ";
+        dumpBitVectorValues(S, S.BBStates[&BB].LiveIn);
+        dbgs() << "\n\tLiveOut: ";
+        dumpBitVectorValues(S, S.BBStates[&BB].LiveOut);
+        dbgs() << "\n\tUnrootedIn: ";
+        dumpBitVectorValues(S, S.BBStates[&BB].UnrootedIn);
+        dbgs() << "\n\tUnrootedOut: ";
+        dumpBitVectorValues(S, S.BBStates[&BB].UnrootedOut);
+        dbgs() << "\n";
+    }
+}
+
+static std::set<Instruction *> DominatingLifetimes(DominatorTree &DT, Instruction *I, Value *V) {
+    // Find all lifetime starts that dominate this call
+    // instruction.
+    std::set<Instruction *> Dead;
+    std::set<Instruction *> LifeTimeStarts;
+    // This could be combined with the visit in the caller or even cached
+    // Don't bother with the additional complexity for now.
+    RecursivelyVisit<IntrinsicInst>([&](Use &VU) {
+        IntrinsicInst *II = cast<IntrinsicInst>(VU.getUser());
+        if (II->getIntrinsicID() == Intrinsic::lifetime_start &&
+            DT.dominates(II, I)) {
+            LifeTimeStarts.insert(II);
+        }
+    }, V);
+    // If one lifetime start dominates another, we can
+    // remove the earlier one.
+    for (auto *LifetimeStart : LifeTimeStarts) {
+        for (auto *LifetimeStart2 : LifeTimeStarts) {
+            if (LifetimeStart == LifetimeStart2 ||
+                Dead.find(LifetimeStart2) != Dead.end())
+                continue;
+            if (DT.dominates(LifetimeStart, LifetimeStart2)) {
+                Dead.insert(LifetimeStart);
+            }
+        }
+    }
+    for (auto *ToDelete : Dead)
+        LifeTimeStarts.erase(ToDelete);
+    return LifeTimeStarts;
+}
+
+State LateLowerGCFrame::LocalScan(Function &F) {
+    State S;
+    auto &DT = getAnalysis<DominatorTreeWrapperPass>().getDomTree();
+    for (BasicBlock &BB : F) {
+        BBState &BBS = S.BBStates[&BB];
+        for (auto it = BB.rbegin(); it != BB.rend(); ++it) {
+            Instruction &I = *it;
+            if (CallInst *CI = dyn_cast<CallInst>(&I)) {
+                if (isa<IntrinsicInst>(CI)) {
+                    // Intrinsics are never GC uses/defs
+                    continue;
+                }
+                MaybeNoteDef(S, BBS, CI, BBS.Safepoints);
+                NoteOperandUses(S, BBS, I);
+                // Note uses via jlcall frames
+                for (Use &U : CI->operands()) {
+                    Value *V = U;
+                    if (isUnionRep(V->getType())) {
+                        NoteUse(S, BBS, V);
+                        continue;
+                    }
+                    if (!V->getType()->isPointerTy())
+                        continue;
+                    if (isSpecialPtr(cast<PointerType>(V->getType())->getElementType())) {
```

Comment:
also, since it's assumed below, might be good to test that it (the eltype) is `T_prjlvalue` instead of being any `isSpecialPtr`